### PR TITLE
Accelerate the do_range_limit_days by advancing multiple months in a …

### PR DESCRIPTION
**Pull request by albymassari**
…single call to the function

Running the body of the new test functions for 1,000,000 times showed that the timelib_add API runs in almost half of the time now (75 seconds instead of 138 seconds)

### CODE REVIEW

File: tm2unixtime.c [hunk 1]
```diff
@@ -104,9 +104,10 @@ static void do_range_limit_days_relative(timelib_sll *base_y, timelib_sll *base_
 static int do_range_limit_days(timelib_sll *y, timelib_sll *m, timelib_sll *d)
 {
 	timelib_sll leapyear;
-	timelib_sll days_this_month;
 	timelib_sll last_month, last_year;
 	timelib_sll days_last_month;
+	int retval = 0;
+	int* days_per_month_current_year;
```
Comment (derickr) [inline tm2unixtime.c:110]: Nit: I usually put the star with the variable name, like: `int *days_per_month_current_year`

File: tm2unixtime.c [hunk 2]
```diff
@@ -117,29 +118,29 @@ static int do_range_limit_days(timelib_sll *y, timelib_sll *m, timelib_sll *d)
...
+	while (*d <= 0 && *m > 0) {
+		last_month = (*m) - 1;
+		if (last_month < 1) {
+			last_month += 12;
+			last_year = (*y) - 1;
+		} else {
+			last_year = (*y);
+		}
+		leapyear = timelib_is_leap(last_year);
+		days_last_month = leapyear ? days_in_month_leap[last_month] : days_in_month[last_month];
```
Comment (derickr) [inline tm2unixtime.c:132]: Would `days_last_month` not be better named as `days_in_previous_months` as it is no longer about just days in a single month?

**Review by derickr — COMMENTED** [review]
I think conceptually this is a good path, with a bunch of minor comments. I would however like to see many more tests, especially ones that cover things around month/year/leap day boundaries, as this is often a place with these calculations go wrong. I will also want to run this with PHP's test suite as it often has these edge cases already tested for.

**Comment by derickr** [comment]
I'll get to it towards the end of the week.

**Comment by derickr** [comment]
Sorry it took so long. But I've now reviewed it, and after fixing memory leaks in the tests (https://github.com/derickr/timelib/commit/ae778a6fe92bd21622bcfc62d1d831919594e8c1) and testing it with PHP's test suite, I have merged this after a rebase. Thanks!
